### PR TITLE
Use plain function pointers for type info accessors.

### PR DESCRIPTION
**Pull request by bbannier**
It turns out that std::function not only adds runtime overhead, but also
that it is significantly slower to compiler than a function pointer.

This commit partially addresses #546.

### CODE REVIEW

File: doc/host-applications.rst [hunk 1]
```diff
@@ -315,5 +315,5 @@ Our examples always passed the full input at once. You don't need to
 do that, Spicy's parsers can process input incrementally as it comes
 in, and return back to the caller to retrieve more. See the source of
 `spicy::Driver::processInput()
-<https://github.com/zeek/spicy/blob/master/spicy/src/rt/driver.cc#L121>`_
+<https://github.com/zeek/spicy/blob/4e3fcb80d19aaa2b9b205134b8e2453348dbaf50/spicy/runtime/src/driver.cc#L144>`_
```
Comment (rsmmr) [inline doc/host-applications.rst:318]: This risks showing outdated code at some point in the future. Not sure what's better: that, or showing wrong line numbers. Hard to catch either way.

File: hilti/runtime/include/type-info.h [hunk 2]
```diff
@@ -180,7 +180,7 @@ class DereferencableType {
 
 private:
     const TypeInfo* _vtype;
-    const Accessor _accessor;
+    Accessor* _accessor;
```
Comment (rsmmr) [inline hilti/runtime/include/type-info.h:183]: Could this remains `const`?

Also, how about turning the `Accessor` typedef into a `pointer to <function>` (i.e., `using Accessor = const void* (*) (const Value& v);`, then we can pass it around as just `Accessor` just like we do further below for `Accessor` types that are tuples.

File: hilti/runtime/include/type-info.h [hunk 3]
```diff
@@ -605,7 +605,7 @@ class Map {
 
     const TypeInfo* _ktype;
     const TypeInfo* _vtype;
-    const Accessor _accessor;
+    Accessor _accessor;
```
Comment (rsmmr) [inline hilti/runtime/include/type-info.h:608]: Same `const` question (and more below).

File: hilti/runtime/include/type-info.h [hunk 4]
```diff
@@ -691,7 +691,7 @@ class MapIterator {
 private:
     const TypeInfo* _ktype;
     const TypeInfo* _vtype;
-    const Accessor _accessor;
+    Accessor* _accessor;
```
Comment (rsmmr) [inline hilti/runtime/include/type-info.h:694]: Should this really be a pointer? `Accessor` is a pair here, no?

**Review by rsmmr — CHANGES_REQUESTED** [review]
Are you seeing a different in compile times?

File: doc/host-applications.rst (hunk 1 above)
Comment (bbannier) [reply doc/host-applications.rst:318]: I changed this to now link to only the file, not the line number.

File: hilti/runtime/include/type-info.h (hunk 4 above)
Comment (bbannier) [reply hilti/runtime/include/type-info.h:694]: Accessor is a function `const Value&` -> `std::pair<const void*, const void*` so this is fine. I changed this to a function pointer type as well like you suggested above.

**Comment by bbannier** [comment]
> Are you seeing a difference in compile times?

I tested with by compiling `hilti/runtime/src/tests/type-info.cc` which includes moderate use of this header and see a significant, but only moderate speedup:

- before: 3.617 s ±  0.018 s
- after: 3.505 s ±  0.033 s

Looking at the time profile a bigger win be possible from getting rid of the `std::variant` use in this header. I'll look into that next.

**Comment by rsmmr** [comment]
Try comparing it for one of the btests, say `spicy/types/unit/sub-unit.spicy`

**Comment by bbannier** [comment]
> Try comparing it for one of the btests, say `spicy/types/unit/sub-unit.spicy`

For that one the difference is bigger:

- before: (11.979 s ± 0.610) s
- after: (9.793 s ± 0.242) s

FTR, I first safed the output to a file, and then benchmarked generating an object file from that source file
```.sh
spicyc -c spicy/types/unit/sub-unit.spicy -o foo.cc
$(spicy-config --cxx --cxxflags --ldflags) -c foo.cc -o foo.o
```

**Comment by rsmmr** [comment]
Great, thanks, that's noticeable.

**Review by rsmmr — APPROVED** [review]

**Pull request merged**